### ISSUE TITLE
missing value treatment

### ISSUE THREAD

**Issue by lineagezyf** [jpmml/sklearn2pmml#190]
Hello, 

This may not be an issue.. but I am trying to understand how correctly output PMML file from LightGBM model when input data has missing. below is the python code I used, data if [from](https://openscoring.io/blog/2019/04/07/converting_sklearn_lightgbm_pipeline_pmml/)

```python
from lightgbm import LGBMClassifier
from sklearn.impute import SimpleImputer
from sklearn.metrics import accuracy_score, precision_score, recall_score, roc_auc_score
from sklearn.preprocessing import LabelBinarizer, LabelEncoder
from sklearn_pandas import DataFrameMapper
from sklearn2pmml import sklearn2pmml
from sklearn2pmml.decoration import CategoricalDomain, ContinuousDomain
from sklearn2pmml.preprocessing import PMMLLabelEncoder
from sklearn2pmml.pipeline import PMMLPipeline

import pandas as pd
from sklearn.pipeline import Pipeline
from sklearn2pmml import make_pmml_pipeline

df = pandas.read_csv("audit-NA.csv")
for col in df.columns:
    if df[col].dtypes == 'object':
        df[col] = df[col].astype('category').cat.codes.astype(float)
    if df[col].dtypes == 'bool':
        df[col] = df[col].astype('category').cat.codes.astype(float)
mapper = DataFrameMapper(
    [(df.columns.values[:-1], ContinuousDomain())]
)

lgbm_config = {
        "n_estimators" : 31,
        "objective" : "binary",
        "random_state" : 42
    }
pipeline_obj = Pipeline([
    ('mapper', mapper),
    ('lgbm', LGBMClassifier(**lgbm_config))
])
pipeline_obj.fit(df.drop('Adjusted',axis=1), df["Adjusted"])
pipeline = make_pmml_pipeline(pipeline_obj, df.drop('Adjusted',axis=1).columns.values)
pipeline.configure(compact = False)
sklearn2pmml(pipeline, "LightGBM_test2.pmml.xml") 
```

**Comment by lineagezyf** [comment]
when I upload this PMML file to IBM watson studio for test, I cannot leave input as blank..

**Comment by vruusmann** [comment]
> when I upload this PMML file to IBM watson studio for test, I cannot leave input as blank..

IBM Watson studio may have issues with 1) interpreting/converting blank values (empty strings?) as missing values or 2) not supporting missing values at all.

Does the prediction work with the [JPMML-Evaluator](https://github.com/jpmml/jpmml-evaluator) library (see the `org.jpmml.evaluator.EvaluationExample` command-line application)?

**Comment by lineagezyf** [comment]
Yes, I was trying to use your evaluator package to verify the PMML file, unfortunately, I have 0 experience in JAVA, and I haven't successfully run your evaluator package yet... I might need to work more on that method though. 

But thanks for your replay, just want to know if the PMML file I generated is supposed to handle missing value correctly.

**Comment by vruusmann** [comment]
> unfortunately, I have 0 experience in JAVA

Java experience is not needed. You're just supposed to download an uber-JAR file from the releases page, and then run it on command line: https://github.com/jpmml/jpmml-evaluator#example-applications

>  just want to know if the PMML file I generated is supposed to handle missing value correctly

All XGBoost and LightGBM estimators should be exported to PMML documents that can handle missing values natively. When dealing with continuous columns, then keep them as `float(NaN)` or `None`, and inform the converter about your decision by setting the `ContinuousDomain.missing_values` attribute to the appropriate value (defaults to `None`). When dealing with categorical columns, then keep them as `None`, optionally set the `CategoricalDomain.mssing_values` attribute, and perform the binarization using either `PMMLLabelBinarizer` (XGBoost) or `PMMLLabelEncoder` (LightGBM).

TLDR: If you managed to train a Scikit-Learn pipeline with sparse training data, and export it to PMML using the `sklearn2pmml.sklearn2pmml` function, then there's a 99% chance that the PMML file is correct. The issue is happening in the IBM Watson studio side - see its documentation about how it expects you to present/encode missing values.
